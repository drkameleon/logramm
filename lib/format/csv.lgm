/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/csv.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;
import core.string;

//================================================
// Module definition
//================================================

Csv :: 
{
	//================================================
	// Functions
	//================================================

	parse(csv) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes csv string and returns corresponding object"
	] => 
	{
		lines = csv.split("\n");
		lines = lines.map("trim($)").filter("$!=\"\"");

		header = lines[0];
		headers = header.split(",").map("trim($)");

		ret = [];

		@ line->lines : 
		{
			line!=lines.first() : {

				parts = line.split(",").map("trim($)").filter("$!=\"\"");

				(count(parts)==count(headers)) :
				{
					row = [:];

					@ i,part->parts : row += [ headers[i] : part ];

					ret += row;
				}
			}
		}

		==> ret;
	}

	generate(object) # [
		.params		: "array",
		.returns	: "string",
		.help		: "Processes input and returns corresponding Json as string"
	] => 
	{
		ret = "";

		headers = object.first().keys().map("trim($)");
		
		ret += headers.join(",") + "\n";

		@ row->object : ret += row.values().join(",") + "\n";

		==> ret;
	}
	
}

