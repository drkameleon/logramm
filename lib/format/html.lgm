/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/html.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

import core.string;

//================================================
// Module definition
//================================================

Html :: 
{

	//================================================
	// Functions
	//================================================

	external parse(markup) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Converts HTML/LGMX template to HTML and returns result"
	] => "html.parse";
	
}
