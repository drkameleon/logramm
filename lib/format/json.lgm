/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/json.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Json :: 
{

	//================================================
	// External D-bindings
	//================================================

	external parse(json) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes json string and returns corresponding object"
	] => "json.parse";

	external generate(object) # [
		.params		: "any",
		.returns	: "string",
		.help		: "Processes input and returns corresponding Json as string"
	] => "json.generate";

}

