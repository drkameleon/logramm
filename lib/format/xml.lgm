/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/xml.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

//================================================
// Module definition
//================================================

Xml :: 
{

	//================================================
	// External D-bindings
	//================================================

	external check(xml) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks an XML document, as string, for well-formedness and returns true if successful"
	] => "xml.check";

}

