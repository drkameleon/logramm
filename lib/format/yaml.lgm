/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/yaml.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Yaml :: 
{

	//================================================
	// External D-bindings
	//================================================

	external parse(yaml) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes YAML string and returns corresponding object"
	] => "yaml.parse";

	external generate(object) # [
		.params		: "any",
		.returns	: "string",
		.help		: "Processes input and returns corresponding YAML as string"
	] => "yaml.generate";

}

