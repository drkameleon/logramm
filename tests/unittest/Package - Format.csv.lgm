import format.csv;

users =
[
	[.username:"drkameleon", .password:"1234567", .age:28],
	[.username:"ozmax",		 .password:"7654321", .age:24]
];

csv = generate(users);

out csv;
out "==============";
out parse(csv);