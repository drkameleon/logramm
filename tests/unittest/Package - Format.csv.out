username,password,age
drkameleon,1234567,28
ozmax,7654321,24

==============
[["username":"drkameleon", "password":"1234567", "age":"28"],["username":"ozmax", "password":"7654321", "age":"24"]]
